### ISSUE TITLE
How to use event stream trough multiple worker

### ISSUE THREAD

**Issue by enzofrnt** [fanout/django-eventstream#151]
HI, how can I set up eventstream to work across multiple worker using redis ?
For the moment I made my own consumer that I start as a thread… That maybe not the best way to do it…

Thanks you.

**Comment by jkarneges** [comment]
I have not tried this but what you describe sounds fine, and is why I included your sentence about this in the readme.

Are you sure `send_event` is actually getting called in the other process? If it is, I can't see how it would be any different than calling it from a view in that other process.

**Comment by enzofrnt** [comment]
> I have not tried this but what you describe sounds fine, and is why I included your sentence about this in the readme.
> 
> 
> 
> Are you sure `send_event` is actually getting called in the other process? If it is, I can't see how it would act any different than calling it from a view in that other process.


I don’t think you fully understood what I meant. When I send an event with Django EventStream via send_event from a process other than the main one, the main process does not receive it and does not send it to the view associated with EventStream.

Currently, I have to send messages to Redis, which are then retrieved by a consumer in the main process. This consumer then sends the events using send_event to send the event to the client.

What I mean is: is there a way to use send_event in another process such that the event is sent to the client, regardless of which Django process is managing the SSE view or sending the event?

I see that Grip may be a solution, but I’m not sure. Maybe this functionality needs to be implemented?

**Comment by jkarneges** [comment]
Ah, yes it would need to be implemented. Currently GRIP is the only way, but a Redis option would be nice. Something that works locally without anything extra (unix socket?) could be nice too for local development.

**Comment by enzofrnt** [comment]
I don’t know what Unix sockets are, but that might be an option. I was thinking that using Redis is a great idea.

**Comment by enzofrnt** [comment]
I just learned what Unix sockets are, and they seem like a pretty great option. I don’t know if I have the required competence to develop something like that, but I would try to help.

**Comment by enzofrnt** [comment]
Where and how we cal start work on this ?

**Comment by enzofrnt** [comment]
Hi @jkarneges,

Here is an example of a working implementation across processes using Redis. It still needs some cleaning up and fixing, and we also need to add a Unix socket version.

If you want to try it, you can start a Django shell using `python manage.py shell` and send messages from this shell using the `send_event` method. 

This configuration is required at the moment to make it work:
```py
EVENTSTREAM_USE_REDIS = True  # Enable or disable Redis Pub/Sub
EVENTSTREAM_REDIS_HOST = 'localhost'
EVENTSTREAM_REDIS_PORT = 6379
EVENTSTREAM_REDIS_DB = 0
```

If we implement this and use a Unix socket as the default solution with Redis as an optional choice, will the GRIP component still be necessary? I'm not sure.

**Comment by jkarneges** [comment]
Nice, that looks like it works the way I'd expect.

GRIP is still a useful alternative that we should keep. It enables scaling connections & messages separately from scaling the server app. It also allows the server app to run as a cloud function/lambda.

**Comment by enzofrnt** [comment]
Ok.
Do you know how we can delete thread for Redis and only use asynchronous task ? Without blocking the execution of Django process ?

**Comment by jkarneges** [comment]
It looks like the redis python client supports async, so maybe it is possible, but I've never used it. Otherwise a daemon thread makes sense to me.

**Comment by enzofrnt** [comment]
Isn't your "heavy" for this ?

**Comment by enzofrnt** [comment]
It’s working without threads! There is just one issue: when sending events from the same process as Django, the SSE receives two events. But this is just a minor adjustment that needs to be made. Next, I will add support for Unix sockets.

**Comment by enzofrnt** [comment]
Hi @jkarneges, I don't know how GRIP works, but won't the use of a Unix socket or Redis cause some messages to be received twice in certain cases?

**Comment by jkarneges** [comment]
Each connection is managed either by django-evenstream's ListenerManager or by the GRIP proxy, but never both. For example, if 5 connections connect directly to django-eventstream and another 5 connect go through a GRIP proxy, the django-eventstream view will be invoked 10 times but only the direct connections will be added to the ListenerManager. When send_event is called, django-eventstream will send directly to its 5 connections, and it will also publish to the proxy in order to reach the other 5. So there shouldn't be any duplicate deliveries, as long as additional mechanisms like a Unix socket or Redis don't themselves cause duplicate deliveries.

To avoid duplicate deliveries with Redis you could omit local delivery when Redis is enabled or ignore messages echoed back.

**Comment by enzofrnt** [comment]
Hi,

I was able to implement the Redis version without major issues, but now I’m encountering a significant problem.

Currently, using Redis or the older version that doesn’t work with overprocess, events can be sent from anywhere, with or without Django running, and everything works fine. However, with the Unix Socket version, it’s not working as expected because if the Socket Server is not initiated, send_event returns an error.

So, is the Unix Socket version really necessary? If we choose to let the issue appear, the error won’t be directly related to a real problem with the socket, but rather with the fact that the Django server is not running at the time the event is sent, even though we have found a way to start the socket asynchronously before.

**Comment by enzofrnt** [comment]
> Each connection is managed either by django-evenstream's ListenerManager or by the GRIP proxy, but never both. For example, if 5 connections connect directly to django-eventstream and another 5 go through a GRIP proxy, the django-eventstream view will be invoked 10 times but only the direct connections will be added to the ListenerManager. When send_event is called, django-eventstream will send directly to its 5 connections, and it will also publish to the proxy in order to reach the other 5. So there shouldn't be any duplicate deliveries, as long as additional mechanisms like a Unix socket or Redis don't themselves cause duplicate deliveries.
> 
> To avoid duplicate deliveries with Redis you could omit local delivery when Redis is enabled or ignore messages echoed back.

So, we use GRIP to prevent SSE from locking or occupying all the workers?
I tried my implementation in production and noticed that a lot of connections can quickly lock up Uvicorn or Daphne.

**Comment by jkarneges** [comment]
GRIP does free up the workers, but lots of connections ought to be possible with ASGI alone since that's the reason async & ASGI exist. Maybe need to tune worker count / thread pools.

**Comment by enzofrnt** [comment]
I think we need to take a look at what you just said because in the current version (to exclude that my work is causing the issue), if I start an ASGI server with about 4 workers, after opening between 3 and 5 connections to the SSE endpoint, the server is likely to become inoperable! I think this is a big issue. Or am I doing something wrong?

**Comment by jkarneges** [comment]
I ran the chat example with `runserver` and manually opened 12 connections with curl and everything seemed fine (messages delivered quickly and cpu usage was low). Obviously a very small test but if you're having trouble at 5 connections then something's definitely wrong.

**Comment by enzofrnt** [comment]
> I ran the chat example with `runserver` and manually opened 12 connections with curl and everything seemed fine (messages delivered quickly and cpu usage was low). Obviously a very small test but if you're having trouble at 5 connections then something's definitely wrong.

Please try with gunicorn or daphne in prod.

**Comment by enzofrnt** [comment]
Also i try with viewset so maybe that come from that

**Comment by enzofrnt** [comment]
Looks like the issue come from my navigator that not able to handle a lot of connection, is that possible ??

**Comment by enzofrnt** [comment]
Hi @jkarneges,
I've identified several architectural concerns in the event handling system of this module, particularly regarding the Listener and event sending mechanisms. Here are the key points and proposed improvements:

1. Architectural Issues:

- The current implementation mixes different event distribution mechanisms

- Events are sometimes sent both to the listener system and GRIP simultaneously

- There's no clear separation between different types of listeners

2. Proposed Solution:

I've started working on a refactoring here that includes:

- An abstract `BaseEventListener` class defining the common interface

- Four distinct listener implementations:

  - `RedisListener`: For Redis-based communication

  - `FileSystemListener`: A new cross-process solution using file-based IPC (Need to be tested on Windows)

  - `InProcessListener`: For single-process communication (Maybe useful in some cases)

  - `GripListener`: For GRIP-based communication

- A improved `ListenerManager` that:

  - Handles the selection of appropriate listener

  - Manages event distribution

  - Provides cleaner separation of concerns

3. Key Questions:

- Would it be acceptable to make the `FileSystemListener` the default solution? 

- Should we separate GRIP event handling from other listener mechanisms to avoid duplicate event sending?

- What are your thoughts on this new architecture that provides better separation of concerns?

The goal is to make the system more maintainable, easier to understand, and provide a default cross-process solution that works without external dependencies. I've already implemented a working version of the `FileSystemListener` that you can find in my [fork](https://github.com/enzofrnt/django-eventstream/tree/default-cross-process-send_event).
Would you be interested in reviewing these changes? I believe they would significantly improve the module's architecture while maintaining compatibility with existing implementations.

**Comment by jkarneges** [comment]
Making the default be file-based IPC sounds great to me, and having a common interface for listeners seems like a reasonable approach

I don't know if GRIP counts as a listener though, since it wouldn't dispatch to local clients like the others would. In any case we'd want it to be possible to use GRIP with one of the other listeners at the same time.

Architecturally, maybe there should be:

* A listener (gets events from somewhere)
* A connection manager (holds client sockets; listens for events from the listener and dispatches to clients)
* A sender (sends to other listeners and/or grip)
